src/librustc_mir/transform/simplify_comparison_integral.rs

```diff
@@ -0,0 +1,252 @@
+use super::{MirPass, MirSource};
+use rustc_ast::ast::{IntTy, UintTy};
+use rustc_middle::{
+    mir::{
+        interpret::{truncate, Scalar},
+        BasicBlock, BinOp, Body, Operand, Place, Rvalue, Statement, StatementKind, TerminatorKind,
+    },
+    ty::{Ty, TyCtxt, TyKind::*},
+};
+use rustc_target::abi::{HasDataLayout, Size};
+
+/// Pass to convert `if` conditions on integrals into switches on the integral.
+/// For an example, it turns something like
+///
+/// ```
+/// _3 = Eq(move _4, const 43i32);
+/// StorageDead(_4);
+/// switchInt(_3) -> [false: bb2, otherwise: bb3];
+/// ```
+///
+/// into:
+///
+/// ```
+/// switchInt(_4) -> [43i32: bb3, otherwise: bb2];
+/// ```
+pub struct SimplifyComparisonIntegral;
+
+impl<'tcx> MirPass<'tcx> for SimplifyComparisonIntegral {
+    fn run_pass(&self, tcx: TyCtxt<'tcx>, source: MirSource<'tcx>, body: &mut Body<'tcx>) {
+        trace!("Running SimplifyComparisonIntegral on {:?}", source);
+
+        let helper = OptimizationFinder { body };
+        let opts = helper.find_optimizations();
+        let mut storage_deads_to_insert = vec![];
+        let mut storage_deads_to_remove: Vec<(usize, BasicBlock)> = vec![];
+        for opt in opts {
+            trace!("SUCCESS: Applying {:?}", opt);
+            // replace terminator with a switchInt that switches on the integer directly
+            let bbs = &mut body.basic_blocks_mut();
+            let bb = &mut bbs[opt.bb_idx];
+            let new_value = scalar_to_u128(opt.branch_value_scalar, opt.branch_value_ty, tcx);
```

Comment:
I believe you don't actually need a type check here, you should be able to invoke `opt.branch_value_scalar.to_bits(Size::from_bytes(16)).unwrap()` and it should never fail unless the compiler has bugs.

Reply:
oh... you're right, we made `to_bits` private on purpose. I still don't think you need your `scalar_to_u128` function. As an invariant of MIR (cc #75440) we can't have wrong values here, so you can just invoke your `to_u128` function and use that value.

Reply:
When changing it to just `opt.branch_value_scalar.to_u128()` i get the following error.
`InterpErrorInfo { kind: scalar size mismatch: expected 16 bytes but got 8 bytes instead`.
I may use the hammer wrong though. Cleaning up the scalar_to_u128 function would be lovely, but unsure how to do that exactly without looking at the type and then delegating to `to_xxx`

Reply:
oh. I did forget about the size check. This is a new case that we haven't really considered this way before. I think in this case, where we know that we need the bits representation for the `values` array it is ok to directly extract them:

```suggestion
            // We only use the bits for the untyped, not length checked `values` field. Thus we are
            // not using any of the convenience wrappers here and directly access the bits.
            let new_value = match opt.branch_value_scalar {
                Scalar::Raw { data, .. } => data,
                Scalar::Ptr(_) => continue,
            };
```

Reply:
Cool, fixed